NOx contamination correction

srweintraub opened the pull request:
As discussed in RITM0055571 and approved by the OS-IPT, I have added an **noxCorrection** parameter to the _def.calc.ntrans_ function so that the quantile regression approach we developed in [this paper](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2022EF003224) can be applied to 2021 and earlier data using the function. The parameter default is set to T as advised by the OS-IPT, so users need to opt out in order to not correct the pre-2022 data. There is also a new field in the all_data output table called noxCorrection that records if the correction was implemented, not implemented, or not applicable (2022 and later data). 

There was a bit of complexity in thinking about how this parameter should interact with the 'drop flags' call for nitrate, I think I've got it working properly but an independent review would be helpful. I have a testing script located here that could be helpful for this purpose: GitHub/definitional-data/wildWestScripts/SWeintraub/neonNTrans_testing.R

I think that perhaps @carenscott or @cklunch could be good reviewers for this PR? I am also happy to rope in someone else from the DP team, just let me know. Thank you.

srweintraub committed:
Update def.calc.ntrans.R

adding a new parameter, noxCorrection, and updating function code and header.

srweintraub committed:
deleted an extraneous bracket

srweintraub committed:
update function header

srweintraub committed:
fixing header details

srweintraub committed:
better handling for when optional parameters not used

srweintraub committed:
increment the package version in Description

srweintraub committed:
column name correction

cklunch commented:
@srweintraub Thanks for the code update! I'm happy to review. I expect I won't have time to look at it until sometime in January, let me know if that timeline would be a problem.

srweintraub commented:
@cklunch that works, I look forward to your feedback in January.

cklunch commented:
@srweintraub This is looking good, everything I've tried runs without error, and the correction looks pretty simple. I'm still a bit confused about the flagging. I ran your test script, and it ran fine, but before I did that, I tried following the instructions in the package readme. Running it that way, with the new code, the dropped_flags table was still generated, but was empty. It had lots of entries if I ran the same data using the old code. Maybe that's expected, but I'm a bit puzzled - does dropConditions override dropAmmoniumFlags and dropNitrateFlags? Maybe just an update to the readme to explain the new options.

srweintraub commented:
@cklunch thanks for the review. By chance could you share the code that you used where you saw this behavior, if it's not in my helper file? or if it is in the helper, point me toward those rows? In looking over the code, it shouldn't make that table if there were no samples that got dropped due to having certain QF values. With the nox correction, it got a bit complex because, even if we flagged them with a QF of 'blanks exceed sample values' I don't drop them as long as the blank-corrected values are now positive. So if all samples in your test set are OK after the correction, then I wouldn't expect to see this table at all...so it's strange that you see it but it's empty. I would like to reproduce that so I can troubleshoot. Thanks.

srweintraub committed:
Update def.calc.ntrans.R

update to remove the 'dropped flags' table if there is nothing in it, because nothing was dropped given the noxCorrection.

srweintraub commented:
@cklunch thanks for sending me your independent testing script. It seems I had omitted a key step, in that if users specified some 'dropFlags' inputs, but then in the end nothing gets dropped because the noxCorrection fixes the data, I was still generating the drop flags output table (it was just empty). Check out my latest commit, I added one extra bit of code to fix this. Can you see if it works as you'd expect it to now? Note that I am deciding to apply the function correction first, then seeing if anything should be dropped...I tried to make this clear in the revised version help file, but if you think that's still confusing (or that I shouldn't be doing it this way), let me know...

cklunch commented:
@srweintraub Yep, I'm getting the same behavior you are now, with the flag table omitted. I think it's all looking good! An update to the readme would still be nice, maybe just a couple more examples of the various drop options.

srweintraub commented:
@cklunch OK good, glad that fixed it. Did you see that I updated the examples in the readme to show one with using drop conditions and another with drop flags and noxCorrection set to false? can you let me know what other examples/combos specifically you would like to see? I'll get those added, thanks.

cklunch commented:
@srweintraub Oh weird, I'm not seeing changes in the readme file, it matches the main branch. Any chance it's on another fork or something?

srweintraub commented:
@cklunch nope. If you go to the 'Files changed' tab of this PR you can see that the .Rd file did change (you can also see the changes in the header of the .R file). And in my R studio, when I pull up the help file for version 0.1.3 I see a revised file, with the noxCorrection parameter included and two chunks of example code showing different parameter configurations. Let me know if you want to share screens and we can try to figure out why you're not seeing the same thing.

srweintraub committed:
readme updates for the package

@cklunch see what you think

cklunch commented:
@srweintraub Documentation updates look great. Thank you!

srweintraub commented:
@cklunch nice, thanks for the careful review. I've been working on DPUG updates as well, about ready to submit a PR for those. Should I merge this pull request or wait for the DPUG to be approved and posted first?

cklunch commented:
@srweintraub Either way is fine, I don't think it matters if these changes go in before the DPUG is updated.

srweintraub commented:
Ok thanks, in that case I shall merge it now.
